design-proposals/klip-57-kafka-headers.md

```diff
@@ -0,0 +1,132 @@
+
+## What is in scope
+* Syntax for accessing headers
+* Support for all query types except for `INSERT` queries
+* Including header columns in a data source's schema
+* Functions to help decode byte strings to primitive types (BYTES to STRING already exists, new functions would be conversions to INT, BIGINT and DOUBLE)
+
+## What is not in scope
+* Writing headers to output topics - we might do this in the future, but that is a separate discussion
+* Advanced functions to help decode byte strings (e.g. functions to deserialize complex serialized data.) - if a user wants to convert something that we don't have a function for, then they can either create their own UDF or use `CREATE FUNCTION` when that becomes available.
+
+## Public APIS
+
+### Syntax
+
+Unlike other pseudocolumns, if a user wants to access headers, they will create a column with the `HEADERS` keyword. This column will be populated with the full list of header keys and values.
+```
+ksql> CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS);
+```
+The type of a header backed column must be exactly `ARRAY<STRUCT<key STRING, value BYTES>>`. An error will be thrown if that is not the case.
+```
+ksql> CREATE STREAM B (my_headers ARRAY<BYTES> HEADERS);
+Error: Columns specified with the HEADERS keyword must be typed as ARRAY<STRUCT<key STRING, value BYTES>>.
+```
+Users can also mark a column with `HEADER(<key>)`, which will populate the column with the last header that matches the key. The data type must be `BYTES`
+```
+ksql> CREATE STREAM B (value BYTES HEADER(<key>));
+```
+If the key is not present in the header, then the value in the column will be `null`.
+
+Sources created with schema inferencing can have header columns, they would just have to be specified. For example, the statement `CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS) WITH (kafka_topic='a', value_format='avro');` will create a stream with columns from Schema Registry, as well as one header column.
+
+### New byte-conversion functions
+
+There will also be new byte conversion functions to help decode header data:
+
+* `INT_FROM_BYTES(bytes, byte_order='big', signed=true)` - The input byte array must be exactly 4 bytes long. Signed conversions will by handled by `ByteBuffer`'s `getInt` method. Unsigned conversions will be manually calculated. If the result does not fit into an INTEGER then the method will throw an error.
```

Comment:
> then the method will throw an error.

What does this imply? Don't think we want to crash/stop a persistent query for this case?

Reply:
When UDFs in ksqlDB throw errors, the error is logged and `null` is written to the output topic. So this won't crash persistent queries.

Reply:
> and null is written to the output topic

For the headers case, we are on the read path. So would we have a header `<k,null>` or drop the header if we cannot parse the header-value? Maybe having `<k,null>` is better to at least indicate to the user that there _is_ a header?

Reply:
That's not a situation that would come up to begin with, since doing something like `FUNCTION(COLUMN_NAME)` creates a new column, rather than writing to `COLUMN_NAME`. So there is no way a UDF would write anything to the header field of an output topic.

Reply:
Added a note with explanation on how error handling will work for UDFs using headers